### ISSUE TITLE
Stake/Latency Based O Selection

### ISSUE THREAD

**Issue by yondonfu** [livepeer/go-livepeer#1215]
## Abstract

This spec outlines a stake/latency based O selection algorithm that Bs can use to select Os to work with.

## Motivation

This design is an extension of #938. It also seeks to address flaws in previous iterations of stake/latency based selection designs which included:

- Increased susceptibility to playback jitter due to a fan out approach that attempted to find the fastest Os in the network. While the fan out approach allowed a B to find Os with lower latency scores than the ones in its current working set, it also increased the variance of latencies since each O used during the fan out process could have very different latencies
- Increased cost overhead incurred by GPU session initialization due to a weighted random selection approach that increased the chance of sporadic segment assignment to Os. Each O incurs a cost overhead for initializing a GPU session for the first segment received in a stream and each session has a fixed timeout. So, if an O does not consistently receive segments, then eventually the session will timeout and the next time the O receives a segment for the same stream it will have incur the cost overhead of GPU session initialization again
- Increased quality variance across segments in a rendition due to a weighted random selection approach that will potentially assign segments of a stream to many different Os. If each O is using different GPUs, then the results produced can vary in quality. In the long term, we may be able to guarantee a certain quality level for each segment in a rendition by [requiring minimum quality scores](https://github.com/livepeer/verification-classifier/issues/71). In the short term, we would benefit from re-using the same GPUs by re-using the same Os

In order to address the flaws in previous iterations, this design's priorities are:

1. Consistent transcode latency. We still want to select Os that produce the fastest result, but more importantly, we want to select Os that *consistently* produce the fastest results
2. Re-use Os as much as possible. Re-using Os allows us to amortize the cost overhead of GPU session initialization across many segments. Additionally, re-using Os can help reduce quality variance in a rendition

## Proposed Solution

The `goodEnoughLatencyScore` parameter is the latency score considered to be "good enough". For example, setting `goodEnoughLatencyScore = 1.0` would mean that B considers real-time transcoding as good enough.

Selection consists of two phases:
- Working set creation
- Working set selection

Working set creation will largely work the same way that it does now - B will call `GetOrchestrator()` on each known O. In on-chain mode B will only send requests to active Os that are not banned (see the section below on dropping and banning). B will use the first N valid responses (where response validity is based on price, PM params, etc.) to form the working set. Using the first N valid responses favors the Os will the fastest initial response times. 

**Dropping and Banning Os**

- If O returns a non-verification related error, it will simply be dropped from the working set [1]
- If O returns a verification related error (this would only happen in on-chain mode), then we evaluate whether O should be banned based on the type of verification related error. Banned Os will not be eligible to be included in the working set. Banning policies might end up being folded into verification policies - see [this comment](https://github.com/livepeer/go-livepeer/issues/1100#issuecomment-558714000) for more details.

For the purposes of selection, B only needs to be aware that there is a ban list that is used to filter out ineligible Os during working set creation. The details of how an O ends up on the ban list can be kept separate from the selection algorithm. The ban list can be implemented separately - the concept is included in this spec just for reader awareness.

[1] We drop Os without further penalization in this case because it is not clear whether the error is ephemeral, a sustained problem with the O, or due to a misbehaving O. One worry here is that the dropped O could be included into the working set again when a refresh occurs and continue to return errors. This might not be that problematic if refresh rates are low (i.e. due to larger working sets with lower min size requirements). But, if it does turn out to be problematic we could consider imposing a time based ban (i.e. 1 minute) to avoid including an O that recently returned an error during a refresh.

**Working set selection**

Currently, the working set consists of a single session list. This design modifies the working set to consist of two session lists:
- `unknownSessList` is an array
    - Tracks Os in the working set that B does not have latency scores for yet
    - When running in on-chain mode, the key of each element in the list is the stake of the O used for the session. The `pop()` operation will pop the O returned by a stake weighted random selection. Each O will be weighted by its stake relative to the cumulative stake of all Os in `unknownSessList`
    - When running in off-chain mode, the key of each element in the list is 0. As a result, the order in which sessions are popped from the list does not matter since all Os will have the same stake weight = 0
- `knownSessList` is a min-heap
    - Tracks Os in the working set that B does have latency scores for
    - The key of each element in the list is the last known latency score of the O used for the session [2]

Working set selection is described by the following pseudocode:

```
func selectSession() *BroadcastSession {
    // Handle working set refresh logic if we are below the min size

    if len(knownSessList) == 0 {
        return pop(unknownSessList)
    }

    if min(knownSessList) > goodEnoughLatencyScore && len(unknownSessList) > 0 {
        return pop(unknownSessList)
    }

    return pop(knownSessList)
}

func completeSession(sess *BroadcastSession) *BroadcastSession {
    ...
    push(knownSessList, sess.lastKnownLatencyScore)
    ...
}
```

In plain English, working set selection consists of the following:
- We have latency scores for known Os, but we don't have latency scores for unknown Os
- Use stake weighted random selection to decide which unknown Os to try and use latency minimizing selection to decide which known Os to use
- Stake defines the probability of O being popped from `unknownSessList`. If an O has X% of the stake within `unknownSessList`, then it has X% probability of being popped next. So, if you can perform well having high stake will be advantageous. But, if you cannot perform well then having high stake might not help you receive more work after the first segment if there are other Os that are faster than you
- We keep selecting from `unknownSessList` until the fastest O in `knownSessList` has a latency score that is considered good enough. It is possible that we run through the full `unknownSessList` in the scenario where no O in `knownSessList` is good enough
- If there is an O in `knownSessList` that is good enough, we will select the fastest O from `knownSessList`
- If we have exhausted all Os in `unknownSessList` without any Os in `knownSessList` that are good enough, we will just use the fastest Os in `knownSessList`. In the case, where `goodEnoughLatencyScore = 1.0`, we may not end up with consistent real-time transcoding, but we'll still end up with the consistently fastest transcoding based on the membership of the working set
- If an O is *consistently* fast then, it will continue to be assigned segments which encourages re-use of Os

[2] For simplicity, we can start with using the last known latency score. A possible improvement, as described in #938, is to use the [EWMA](https://en.wikipedia.org/wiki/EWMA_chart) latency score. If/when we decide to use the EWMA we will need to decide on the two parameters required by the EWMA formula

**Example**

![Orch Selection](https://user-images.githubusercontent.com/5933273/69648100-a3c48600-1038-11ea-802a-4e7c0ed70a94.jpg)

## Implementation Tasks and Considerations

- Cache the stake of each active O at the beginning of a round
- Allow `BroadcastSessionsManager` to use different strategies for selecting the next session by injecting a `BroadcastSessionsSelector` interface as a dependency
- Move the current LIFO selection strategy into a `LIFOSelector` type that implements `BroadcastSessionsSelector`
- Implement `MinLSSelector` Part 1
    - Add a `LatencyScore` field to `BroadcastSession`
    - Add a `BroadcastSession` min-heap implementation that is sorted by `LatencyScore`
    - Use the min-heap as `knownSessList`
- Implement `MinLSSelector` Part 2
    - Use a slice as `unknownSessList`
    - Add a method that returns a session using weighted random selection with the weight of each session being the stake of the session's O. The stake can be read from the DB
- Implement `MinLSSelector` Part 3
   - Add the logic for popping from `unknownSessList` vs. `knownSessList` as described in "Proposed Solution"
   - Make sure `MinLSSelector` implements `BroadcastSessionsSelector`
   - Use `MinLSSelector` in `BroadcastSessionsManager`
- Return the transcode duration from `SubmitSegment()`
- Instead of mutating the `LatencyScore` field in a session, when a response is received with a transcode duration, calculate the latency score, copy the old session's fields into a new session and set the new latency score. The new session will be added back to `BroadcastSessionsManager`. Related #961
- Look into using EWMA latency score instead of last known latency score in a session

## Testing Tasks and Considerations

- Check that the impl follows the specified algorithm
- Check that if all Os in the working set do not meet the `goodEnoughLatencyScore` threshold that the B runs through the entirety of `unknownSessList`
- Check that for a fixed set of Os that are a part of the working set, each O is the first to be tried from `unknownSessList` X% of the time if it has X% of the stake in the working set
- Check that the fastest O in the working set is used as long as either `len(unknownSessList) = 0` or `min(knownSessList) <= goodEnoughLatencyScore`
- Run a test where the stake for Os follows a power law distribution and only low stake Os meet the `goodEnoughLatencyScore` threshold
- Run a test where the stake for Os follows a power law distribution and all Os meet the `goodEnoughLatencyScore` threshold
- Run a test where the stake for Os follows a power law distribution and no Os meet the `goodEnoughLatencyScore` threshold 
- Run a test where the fastest O is heavily utilized and its latency score increases such that it is no longer the fastest in the working set
- If we implement EWMA latency score, try different values for the EWMA parameters

## Known Unknowns

- Reasonable to start with `goodEnoughLatencyScore = 1.0`?
- Will dropping without a time based exclusion from the working set for non-verification related errors negatively impact selection if Os are included into the working set soon after being dropped?

## Alternatives

**Single Session List**

Instead of maintaining two separate lists for the working set, we could use a single list either implemented as a stable min-heap or a sorted array. The stability property is required so that we can select the O from the list with the lowest latency score while also defining the initial ordering of the list based on stake with initial latency scores set to some constant value (for example, if the initial latency score values were set to 0, then we would go through all the Os in the list first). A stable min-heap seems to be [possible](https://cstheory.stackexchange.com/questions/593/is-there-a-stable-heap), but we would need to investigate if there is a readily available implementation. A sorted array would require re-sorting whenever an O is added back to the list. Maintaining two lists seems more straightforward IMO.

**Stake Descending Order**

Instead of popping from `unknownSessList` using a stake weighted random selection, we could implement `unknownSessList` as a max-heap and:
    - When running in on-chain mode, the key of each element in the list is the stake of the O used for the session. As a result, `pop()` operation will always return the O with the greatest stake in the list
    - When running in off-chain mode, the key of each element in the list is 0. As a result, the order in which sessions are popped from the list does not matter. If using a heap structure to maintain `unknownSessList` when we do not care about ordering in off-chain mode feels ineffective, we can also use a vanilla array when running in off-chain mode

A downside of deterministically selecting from `unknownSessList` in descending order of stake is that if the O with the most stake is good enough then other Os on `unknownSessList` will never receive any segments. Even if lower stake Os are faster, they would never be tried because the higher stake O is used first and is already good enough. A stake weighted random selection at the very least gives Os a probability to be tried that is proportional to the amount of stake that they have. Another way to think about this is that a good enough O in `unknownSessList` can "win" the first segment of a stream with a probability proportional to their stake and they will keep receiving subsequent segments for as long as they can remain good enough.

## Additional Context

This design could be extended to utilize redundancy to speed up the process of finding the fastest Os in the working set. Instead, of selecting a single O at a time, we could select N Os at a time. We would select `m` Os from `unknownSessList` and `n` Os from `knownSessList` such that `m + n = N`. We would always use the fastest result from the N Os selected for the playlist, but the redundant transcodes would also push Os onto `knownSessList` at a faster rate. As long as `knownSessList` is not empty, `n >= 1` because we want to always use the fastest known O in the set of N Os selected.

**Comment by dob** [comment]
I'd be in favor of exploring selection from `unknownSessList` not deterministically in decreasing order of stake, but instead randomly in proportion to stake weight. This way, for example, X% active stake would give you first chance to win the work X% of the time, guaranteeing everyone gets a shot to win the work in proportion to their stake. As stated in the initial proposal, the node with the highest stake weight would get first chance to win the work near 100% of the time, and nodes with lower stake weights would have almost minuscule probability of every seeing a job.

**Comment by yondonfu** [comment]
I'd be in favor using stake weighted random selection to pop from `unknownSessList` [1]. I like the property of giving a good enough O on `unknownSessList` the probability of "winning" the first segment of a stream (and they would keep being assigned segments for as long as they are good enough) proportional to their stake. I've updated the OP to use stake weighted random selection for `unknownSessList` and moved the description for the stake descending order selection for `unknownSessList` to the alternatives section.

[1] For clarification, this would *just* be used to pop from `unknownSessList` and we would still minimize latency scores with `knownSessList`.

**Comment by yondonfu** [comment]
Updated the OP to follow the new spec template.

**Comment by yondonfu** [comment]
Closed by #1241
